daft/udf.py

```diff
@@ -100,42 +106,85 @@ def run_udf(
         if name in pyvalues:
             args.append(pyvalues[name])
         else:
-            args.append(evaluated_expressions[function_parameter_name_to_index[name]])
+            # we fill in expressions later
+            expr_name_to_args_index[name] = len(args)
+            args.append(None)
 
     kwargs = {}
+    kwarg_exprs = []  # list of names of exprs in kwargs
     for name in kwarg_keys:
         assert name in pyvalues or name in function_parameter_name_to_index
         if name in pyvalues:
             kwargs[name] = pyvalues[name]
         else:
-            kwargs[name] = evaluated_expressions[function_parameter_name_to_index[name]]
+            kwarg_exprs.append(name)
+
+    # Fill in args and kwargs with sliced versions of the series
+    def set_args_to_slice(start: int, end: int):
+        for name, ind in expr_name_to_args_index.items():
+            expr = evaluated_expressions[function_parameter_name_to_index[name]]
+            if start == 0 and end >= len(expr):
+                # avoid copying if we are taking entire thing
+                args[ind] = expr
+            else:
+                args[ind] = expr.slice(start, end)
+        for name in kwarg_exprs:
+            expr = evaluated_expressions[function_parameter_name_to_index[name]]
+            if start == 0 and end >= len(expr):
+                kwargs[name] = expr
+            else:
+                kwargs[name] = expr.slice(start, end)
+
+    if batch_size is None:
+        set_args_to_slice(0, len(evaluated_expressions[0]))
```

Comment:
Instead of having `set_args_to_slice`, how about we do `get_args_for_slice`?

```
args, kwargs = get_args_for_slice(0, N)
...
func(*args, **kwargs)
```

I think it's fine to construct a new `args` and `kwargs` on the fly.